client/checkout/index.js

```diff
@@ -500,8 +503,8 @@ jQuery( function ( $ ) {
 	maybeShowAuthenticationModal();
 
 	// Handle hash change - used when authenticating payment with SCA on checkout page.
-	window.addEventListener( 'hashchange', function ( event ) {
-		if ( 0 < event.newURL.indexOf( '#wcpay-confirm-pi' ) ) {
+	window.addEventListener( 'hashchange', function () {
+		if ( 0 < window.location.hash.startsWith( '#wcpay-confirm-pi' ) ) {
```

Comment:
This is probably not a concern, but the previous code would match `#wcpay-confirm-pi` even if it wasn't the first string in the hash (e.g. `#somestring#wcpay-confirm-pi`), while the new one wouldn't.

It's unlikely that we'll need to add two parts to the hash, but do you think it would be worth it to search for `#wcpay-confirm-pi` in the entire hash instead of just its beginning?

If we stick to `startsWith`, I think it would be better to remove the `0 < ` comparison, as `startsWith` returns a boolean instead of the index of the occurrence.

Reply:
> the previous code would match #wcpay-confirm-pi even if it wasn't the first string in the hash (e.g. #somestring#wcpay-confirm-pi), while the new one wouldn't.

That's not a problem. [Here](https://github.com/Automattic/woocommerce-payments/blob/af28be94392cf414f203f15271535155525f4c6a/assets/js/wcpay-checkout.js#L396-L398) the URL to "redirect to" is built, there's no way for that URL to contain anything other than `#wcpay-confirm-pi...` in the hash. Also, [here](https://github.com/Automattic/woocommerce-payments/blob/af28be94392cf414f203f15271535155525f4c6a/assets/js/wcpay-checkout.js#L396-L398) there's a RegExp that's matching the whole hash, so the assumption that there's nothing else in the hash is already backed in.

> If we stick to startsWith, I think it would be better to remove the 0 < comparison, as startsWith returns a boolean instead of the index of the occurrence.

Oh, good catch! Of course it still works, because JavaScript 😂  Fixed in https://github.com/Automattic/woocommerce-payments/pull/902/commits/5be2767f7e3fbb657839564e212b0e8bb69b2cb5

Reply:
> That's not a problem. Here the URL to "redirect to" is built, there's no way for that URL to contain anything other than #wcpay-confirm-pi... in the hash. Also, here there's a RegExp that's matching the whole hash, so the assumption that there's nothing else in the hash is already backed in.

Makes sense. Thanks for clarifying it!

> Of course it still works, because JavaScript 😂

😆  - [because JavaScript](https://img.devrant.com/devrant/rant/r_1795923_nJRoc.jpg)

Reply:
hahahahaha that meme is great!